### PR TITLE
discovery: Use map for environment variables

### PR DESCRIPTION
**Pull request by vitkyrka**
### What does this PR do?

Convert the environment variables list to a map early on and use that instead of traversing the list repeatedly in many different functions.

### Motivation

https://datadoghq.atlassian.net/browse/USMON-1169

### CODE REVIEW

**Comment by guyarb** [comment]
Can you share any benchmark result backing up this change?

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go [hunk 1]
```diff
@@ -199,6 +199,22 @@ type parsingContext struct {
 	netNsInfo map[uint32]*namespaceInfo
 }
 
+// envsToMap splits a list of strings containing environment variables of the
+// format NAME=VAL to a map.
+func envsToMap(envs []string) map[string]string {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/impl_linux.go:204]: nit:

```suggestion
func envsToMap(envs ...string) map[string]string {
```

File: pkg/collector/corechecks/servicediscovery/language/language.go [hunk 2]
```diff
@@ -89,16 +89,16 @@ func (pi ProcessInfo) FileReader() (io.ReadCloser, bool) {
...
-		if key, val, _ := strings.Cut(env, "="); key == "PATH" {
-			paths := strings.Split(val, ":")
-			for _, path := range paths {
-				if r, found := findFile(path + string(os.PathSeparator) + fileName); found {
-					return r, true
-				}
+	if val, ok := pi.Envs["PATH"]; ok {
+		paths := strings.Split(val, ":")
+		for _, path := range paths {
+			if r, found := findFile(path + string(os.PathSeparator) + fileName); found {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/language/language.go:95]: nit: (can be moved to a different PR as it is not relevant to the change you're introducing)
```suggestion
			if r, found := findFile(filepath.Join(path, fileName)); found {
```

**Review by guyarb — APPROVED** [review]

File: pkg/collector/corechecks/servicediscovery/apm/detect_nix_test.go [hunk 3]
```diff
@@ -19,25 +19,22 @@ func Test_javaDetector(t *testing.T) {
 	data := []struct {
 		name   string
 		args   []string
-		envs   []string
```
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/apm/detect_nix_test.go:22]: I would recommend leaving this field in the anonymous testing struct (converted to map[string]string to match the new API).

File: pkg/collector/corechecks/servicediscovery/usm/service.go [hunk 4]
```diff
@@ -91,18 +91,14 @@ func NewDetectionContext(logger *zap.Logger, args []string, envs []string, fs fs
...
 
-func extractEnvVar(envs []string, name string) (string, bool) {
-	value := ""
-	prefix := name + "="
-	for _, v := range envs {
-		if strings.HasPrefix(v, prefix) {
-			_, value, _ = strings.Cut(v, "=")
-			break
-		}
+func extractEnvVar(envs map[string]string, name string) (string, bool) {
```
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/usm/service.go:98]: should probably remove this function entirely.

File: pkg/collector/corechecks/servicediscovery/usm/service.go [hunk 5]
```diff
@@ -140,25 +136,22 @@ var binsWithContext = map[string]detectorCreatorFn{
...
-			values := strings.Split(v[len("DD_INJECTION_ENABLED="):], ",")
-			for _, v := range values {
-				if v == "service_name" {
-					fromDDService = false
-					break outer
-				}
+	if env, ok := envs["DD_INJECTION_ENABLED"]; ok {
+		values := strings.Split(env, ",")
+		for _, v := range values {
+			if v == "service_name" {
```
[the remark this answers is not in the document]
Comment (jonbodner) [reply pkg/collector/corechecks/servicediscovery/usm/service.go:144]: This is a bad idea. It introduces a hidden requirement that "service_name" cannot be a substring of any other value in the list of values for DD_INJECTION_ENABLED. That is true today, but might not be in the future.

**Review by jonbodner — APPROVED** [review]
I'm not sure that this change is worth it. The average number of env vars on a process is probably on the order of 10, and a linear search through a slice might be faster than a map lookup, given the overhead of calculating the hash. But we are likely talking about 1 or 2 milliseconds, so it's not discernible. If you think this makes the code more readable, we should do it.

File: pkg/collector/corechecks/servicediscovery/usm/service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/usm/service.go:144]: The code should never have hidden requirements

But here is a zero allocation code, that is also optimal in runtime, and preserves the "hidden requirements"

```go
func elementExists(fullString, subString string) bool {
	// Start and end indices of the current element
	start := 0

	for i := 0; i <= len(fullString); i++ {
		// Check if we've reached the end of a service name
		if i == len(fullString) || fullString[i] == ',' {
			// Compare the current element with the serviceName
			if fullString[start:i] == subString {
				return true
			}
			// Move the start index to the next character after the comma
			start = i + 1
		}
	}

	return false
}
```

Benchmarks

```go
const (
	testString = "value1,value2,value3,value4,value5,sub_service_name,value6,value7,service_name,value8,value9,value10,value11,value12,value13,value14,value15,value16,value17,value18,value19,value20"
)

func BenchmarkSplit(b *testing.B) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		values := strings.Split(testString, ",")
		for _, v := range values {
			if v == "service_name" {
				break
			}
		}
	}
}

func BenchmarkFind(b *testing.B) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		elementExists(testString, "service_name")
	}
}
```

results
```go
BenchmarkSplit
BenchmarkSplit-16    	 6246964	       198.3 ns/op	     352 B/op	       1 allocs/op
BenchmarkFind
BenchmarkFind-16     	19673295	        62.79 ns/op	       0 B/op	       0 allocs/op
```

File: pkg/collector/corechecks/servicediscovery/apm/detect_nix_test.go (hunk 3 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/apm/detect_nix_test.go:22]: Done.

File: pkg/collector/corechecks/servicediscovery/language/language.go (hunk 2 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/language/language.go:95]: Leaving this change out of this PR since it's not strictly related to its purpose.

File: pkg/collector/corechecks/servicediscovery/module/impl_linux.go (hunk 1 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/module/impl_linux.go]: Done.

File: pkg/collector/corechecks/servicediscovery/usm/service.go (hunk 4 above)
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/usm/service.go:98]: Leaving this change out of this PR since it's not strictly related to its purpose.

File: pkg/collector/corechecks/servicediscovery/usm/service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/usm/service.go:144]: Leaving this change out of this PR since it's not strictly related to its purpose.

**Comment by vitkyrka** [comment]
> Can you share any benchmark result backing up this change?

There are benchmarks for slice vs map for strings here: https://darkcoding.net/software/go-slice-search-vs-map-lookup/. The blog is some years old but running the benchmarks yields comparable results.
